typing/typecore.ml

```diff
@@ -4207,7 +4208,12 @@ and type_construct env loc lid sarg ty_expected_explained attrs =
     List.map2 (fun e (t,t0) -> type_argument ~recarg env e t t0) sargs
       (List.combine ty_args ty_args0) in
   if constr.cstr_private = Private then
-    raise(Error(loc, env, Private_type ty_res));
+    begin match constr.cstr_tag with
+    | Cstr_extension _ ->
+        raise(Error(loc, env, Private_constructor_extension (constr, ty_res)))
+    | _ ->
```

Comment:
I would prefer an exhaustive pattern-match here.

Another option would to have a single exception `Private_constructor(constr, ty_res)` and decide how to formulate the error message by looking at `constr.cstr_tag` at the handling site. I have no preference and the current approach seems fine.

Reply:
I prefer the current implementation over trying to reinterpret the context when printing the error message.
I think @Drup (who discussed this change on IRC) will also agree with me.

Reply:
Yes and no.
1. I'm ... actually not convinced by the implementation. It basically approximates "extension & private type => private constructor". Is that really correct ? So far, all private constructors are extensions, but it feels like those two properties should be distinct.
2. If we can decouple "private constructor" from "extension constructor", then I agree with @gasche's remark. In any case, we definitely need a dedicate error (ie., not `Private_type`).

Reply:
@Drup actually, line 4210 checks whether the `private` field of the **constructor** (and not the type) is indeed `Private` as far as I understand. So this should check "extension constructor & private constructor".

Reply:
The assumption is actually the reverse of what Drup said: `private constructor & ¬extension ⇒ private type`.
Before this PR it was just `private constructor ⇒ private type`.

The current PR is actually more correct than what was there before, so I think it's fine.
The assumption might be invalidated in the future, but that's another matter.

Reply:
Right. I feel like we could try to do something slightly less brittle already, but if you think it's better left for later, that's fair.

Reply:
I chatted offline with @Drup, what he had in mind was doing something:
```
if cstr is private then (
  if type is private then
    raise Private_type
  else
    raise Private_constructor
)
```
but that is actually incorrect for extensible types: a private extensible type just mean that you can't add constructors to it.
So if you have:
```ocaml
module M : sig
  type t = private ..
  type t += A
  type t += private B
end

let x = M.B
```
you should really error because the constructor `B` is private (as this PR does), not because `t` is private.

Reply:
Btw, the underscore that the initial remark from @gasche complained about has been replaced by an exhaustive or-pattern. So I'm marking this conversation as resolved.